### PR TITLE
Restore visual_scale support for nodeboxes (and allfaces)

### LINKED ISSUE #8400: visual_scale no longer applies to nodeboxes
##### Issue type
 - Bug report
 
##### Minetest version
5.0.0

##### OS / Hardware
Operating system: kubuntu 19.04
CPU: amd64
GPU model: nvidia ion.2
OpenGL version: 3.3

##### Summary
In minetest 0.4.16 and 0.4.17 it was possible to use visual_scale to resize a nodebox, it seems this feature has been lost in 5.0.0.

##### Steps to reproduce
Create a full node with drawtype "nodebox" and add visual_scale = 0.5 to the mix.

### PR DESCRIPTION
**Pull request by numberZero**
Fix #8400 

`master` | Larger nodebox | This PR
---|---|---
screenshot_20200521_155301 | screenshot_20200521_155236 | screenshot_20200521_160831

## To do

This PR is Ready for Review.

## How to test

Place any node with nodebox (or allfaces) drawtype and visual_scale≠1, like
```lua
minetest.register_node("mese_crystals:mese_pack", {
	description = "Mese pack",
	tiles = {"default_mese_block.png"},
	drawtype = "nodebox",
	node_box = {
		type = "fixed",
		fixed = {-0.4, -0.4, -0.1, 0.4, 0.4, 0.1},
	},
	paramtype = "light",
	paramtype2 = "facedir",
	color = color,
	groups = {cracky = 1},
	use_texture_alpha = true,
	sounds = sounds,
	light_source = 9,
	visual_scale = 2.0,
})
```

Note: glasslike_framed and fencelike will probably look weird with visual scale applied.

### CODE REVIEW

**Comment by numberZero** [comment]
Note: this should not, but may, affect world-aligned textures. Combining visual_scale with these in a single node is not supported.

File: src/client/content_mapblock.cpp [hunk 1]
```diff
@@ -366,13 +366,22 @@ void MapblockMeshGenerator::generateCuboidTextureCoords(const aabb3f &box, f32 *
...
 	f32 dx1 = box.MinEdge.X;
 	f32 dy1 = box.MinEdge.Y;
 	f32 dz1 = box.MinEdge.Z;
 	f32 dx2 = box.MaxEdge.X;
 	f32 dy2 = box.MaxEdge.Y;
 	f32 dz2 = box.MaxEdge.Z;
+	if (scale) {
+		if (!txc) { // generate texture coords before scaling
+			generateCuboidTextureCoords(box, texture_coord_buf);
+			txc = texture_coord_buf;
```
Comment (sfan5) [inline src/client/content_mapblock.cpp:380]: I don't think this special casing is needed, just move the `if (!txc) {` before the `if (scale) {`

**Comment by sfan5** [comment]
Which drawtypes `visual_scale` applies to is documented somewhere in lua_api.txt, please update that too.

File: src/client/content_mapblock.cpp (hunk 1 above)
Comment (numberZero) [reply src/client/content_mapblock.cpp:380]: That would break world-aligned textures

Comment (sfan5) [reply src/client/content_mapblock.cpp:380]: Hm, okay.

**Comment by KaylebJay** [comment]
This will be quite useful. I remember running into this bug a few months ago. Thanks :)

**Comment by numberZero** [comment]
> Which drawtypes visual_scale applies to is documented somewhere in lua_api.txt, please update that too.

Done.

**Pull request merged**